### PR TITLE
[fix] Add capability with DLManagedTensorVersioned

### PR DESCRIPTION
**Pull request by icfaust**
## Description

This PR updates dlpack support to include modern capabilities including:

 - Output conversion of oneDAL tables to DLManagedTensorVersioned objects, which is now default in NumPy, and will be in the next PyTorch release
 - Add support for keywords made available in the array API 2023.12 standard
 - Modify default behavior to make output data not read only (will be modified for all output table conversions in the future, but this makes all conversions consistent)
 - Allow for the ability to convert to host when requested by the ```dl_device``` keyword argument.
 - Small refactor to allow for re-use of code
 - use ```is_none``` call instead of ```.is(py::none())``` to simplify code
 - update dlpack.h from v1.0 to v1.1
 
No performance benchmarks necessary, tests added to cover new conditions.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes or created a separate PR with update and provided its number in the description, if necessary.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have measured performance for affected algorithms using [scikit-learn_bench](https://github.com/IntelPython/scikit-learn_bench) and provided at least summary table with measured data, if performance change is expected.
- [x] I have provided justification why performance has changed or why changes are not expected.
- [x] I have provided justification why quality metrics have changed or why changes are not expected.
- [x] I have extended benchmarking suite and provided corresponding scikit-learn_bench PR if new measurable functionality was introduced in this PR.

### CODE REVIEW

File: onedal/datatypes/_data_conversion.py [hunk 1]
```diff
@@ -111,7 +111,9 @@ def return_type_constructor(array):
                 else xp.asarray(backend.from_table(x), device=device)
             )
     elif hasattr(array, "__array_namespace__"):
-        func = array.__array_namespace__().from_dlpack
+        xp = array.__array_namespace__()
+        device = array.device
+        func = lambda inp: xp.from_dlpack(inp, device=device)
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/datatypes/_data_conversion.py:117]: Can be addressed as a follow up if considered to be an expected  failure mode as we would need to modify more in this function in that case.

File: onedal/datatypes/dlpack/data_conversion.cpp [hunk 2]
```diff
@@ -178,77 +179,131 @@ DLDevice get_dlpack_device(const dal::table& input) {
...
+py::capsule construct_dlpack(const dal::table& input,
+                             py::object max_version,
+                             py::object dl_device,
+                             py::object copyobj) {
     // check table type and expose oneDAL array
     if (input.get_kind() != dal::homogen_table::kind())
-        throw pybind11::type_error("Unsupported table type for dlpack conversion");
+        throw py::type_error("Unsupported table type for dlpack conversion");
+
+    bool copy = !copyobj.is(py::bool_(false));
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp:253]: It is intended behavior, will be swapped in a future PR.

File: onedal/datatypes/_data_conversion.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/datatypes/_data_conversion.py:117]: no

File: onedal/datatypes/dlpack/data_conversion.cpp [hunk 3]
```diff
@@ -178,77 +179,131 @@ DLDevice get_dlpack_device(const dal::table& input) {
...
-    dlm->deleter = [](struct DLManagedTensor* self) -> void {
-        auto stored_array = static_cast<dal::array<byte_t>*>(self->manager_ctx);
-        if (stored_array) {
-            delete stored_array;
+    // oneDAL tables are by definition immutable and must be made mutable via a copy.
+    if (copy)
+        array.need_mutable_data();
+
+    if (max_version.is_none() ||
+        max_version.cast<py::tuple>()[0].cast<int>() < DLPACK_MAJOR_VERSION) {
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp:270]: The current logic is sufficient.

File: onedal/datatypes/common.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+*******************************************************************************/
+
+#include "oneapi/dal/array.hpp"
+
+namespace oneapi::dal::python {
+
+template <typename T>
+dal::array<T> transfer_to_host(const dal::array<T> &array) {
+#ifdef ONEDAL_DATA_PARALLEL
+    auto opt_queue = array.get_queue();
```
Comment (david-cortes-intel) [inline onedal/datatypes/common.hpp:24]: Will GPU arrays always necessarily have an associated queue?

File: onedal/datatypes/tests/test_data.py [hunk 5]
```diff
@@ -616,3 +616,44 @@ def test_table___dlpack__(dataframe, queue, order, data_shape, dtype):
...
+
+
+@pytest.mark.skipif(
+    not hasattr(np, "from_dlpack"), reason="no dlpack support in installed numpy"
+)
+@pytest.mark.parametrize("dataframe,queue", get_dataframes_and_queues("dpctl", "cpu,gpu"))
+@pytest.mark.parametrize("order", ["F", "C"])
+@pytest.mark.parametrize("data_shape", data_shapes)
+@pytest.mark.parametrize("dtype", [np.float32, np.float64, np.int32, np.int64])
+def test_table_cvt_to_host_dlpack(dataframe, queue, order, data_shape, dtype):
```
Comment (david-cortes-intel) [inline onedal/datatypes/tests/test_data.py]: Does 'cvt' mean 'convert'? If so, perhaps the longer name could be used.

File: onedal/datatypes/dlpack/data_conversion.cpp [hunk 6]
```diff
@@ -178,77 +179,131 @@ DLDevice get_dlpack_device(const dal::table& input) {
...
+                                                     homogen_input.get_column_count(),
+                                                     homogen_input.get_metadata().get_data_type(0),
+                                                     homogen_input.get_data_layout(),
+                                                     copy);
+
+        // create capsule
+        capsule = py::capsule(static_cast<void*>(dlm), "dltensor", free_capsule);
+    }
+    else {
+        // assume RVO
```
Comment (david-cortes-intel) [inline onedal/datatypes/dlpack/data_conversion.cpp]: What does 'RVO' mean?

File: onedal/datatypes/dlpack/data_conversion.cpp [hunk 7]
```diff
@@ -178,77 +179,131 @@ DLDevice get_dlpack_device(const dal::table& input) {
...
+
+        // by definition for oneDAL tables, unless a copy of the array is made, it is readonly.
+        if (copy) {
+            dlmv->flags = DLPACK_FLAG_BITMASK_IS_COPIED;
         }
-        delete[] self->dl_tensor.shape;
-        delete self;
-    };
+        else {
+            dlmv->flags = DLPACK_FLAG_BITMASK_READ_ONLY;
```
Comment (david-cortes-intel) [inline onedal/datatypes/dlpack/data_conversion.cpp:300]: Are there tests checking for this code branch and property?

File: onedal/datatypes/dlpack/data_conversion.cpp [hunk 8]
```diff
@@ -178,77 +179,131 @@ DLDevice get_dlpack_device(const dal::table& input) {
...
-    if (PyCapsule_IsValid(cap, "dltensor")) {
-        dlm = static_cast<DLManagedTensor*>(PyCapsule_GetPointer(cap, "dltensor"));
-        if (dlm->deleter) {
-            dlm->deleter(dlm);
+static inline void move_dlpack_data(dal::array<byte_t>& array, py::tuple dl_device, bool copy) {
+    DLDevice requested{ dl_device[0].cast<DLDeviceType>(), dl_device[1].cast<std::int32_t>() };
+#ifdef ONEDAL_DATA_PARALLEL
+    DLDevice current = get_dlpack_device(array);
+    // only allow transfers to host, dlpack implementation does not provide sufficient
+    // fine-grained control for SYCL devices.
```
Comment (david-cortes-intel) [inline onedal/datatypes/dlpack/data_conversion.cpp:230]: Is this because dlpack doesn't include the queue object?

Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp:230]: That is correct!

EDIT: To be fully explicit, since oneDAL being context agnostic via the queues (which limits context handling and other specific things), it would be a burden and require some development to create an allocation on another device without assuming the default context, which we wouldn't know about from the dl_device info given.  Ideally the user should handle movement from sycl devices to other sycl devices on their own, since we in oneDAL only handle from host to device and back. There some tool like dpctl may be better, or using the native conversions in the external library itself.

File: onedal/datatypes/dlpack/data_conversion.cpp (hunk 6 above)
Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp]: My bad, a small note if someone looks at why I am not creating the object and then passing it in by reference. https://en.wikipedia.org/wiki/Copy_elision

File: onedal/datatypes/common.hpp (hunk 4 above)
Comment (icfaust) [reply onedal/datatypes/common.hpp:24]: get_queue returns a ```std::optional``` when used in a dpc build which its own pain. Good catch.

File: onedal/datatypes/tests/test_data.py (hunk 5 above)
Comment (icfaust) [reply onedal/datatypes/tests/test_data.py]: Sounds good. Will change.

File: onedal/datatypes/dlpack/data_conversion.cpp (hunk 7 above)
Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp:300]: Added.

File: onedal/datatypes/dlpack/data_conversion.cpp (hunk 6 above)
Comment (david-cortes-intel) [reply onedal/datatypes/dlpack/data_conversion.cpp]: Could you add a comment in the code?

File: onedal/datatypes/common.hpp (hunk 4 above)
Comment (david-cortes-intel) [reply onedal/datatypes/common.hpp:24]: Then wouldn't this function produce something that will later on cause errors if the array is on GPU but has no queue?

Comment (icfaust) [reply onedal/datatypes/common.hpp:24]: That would never be the case, oneDAL gpu data arrays always have a queue. @Vika-F @Alexandr-Solovev can comment more

File: onedal/datatypes/dlpack/data_conversion.cpp (hunk 6 above)
Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp]: To be able to fulfill your request, please let me know in more detail what beyond what is already written in the comment is necessary. @david-cortes-intel

Comment (david-cortes-intel) [reply onedal/datatypes/dlpack/data_conversion.cpp]: It says "assume RVO". What does that mean? Why can it be assumed in this 'if' branch but not the other?

File: onedal/datatypes/common.hpp [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    if (opt_queue.has_value()) {
+        auto device = opt_queue->get_device();
+        if (!device.is_cpu()) {
+            const auto *device_data = array.get_data();
+
+            auto memory_kind = sycl::get_pointer_type(device_data, opt_queue->get_context());
+            if (memory_kind == sycl::usm::alloc::unknown) {
+                throw std::runtime_error("[convert_to_numpy] Unknown memory type");
+            }
+            if (memory_kind == sycl::usm::alloc::device) {
```
Comment (Alexandr-Solovev) [inline onedal/datatypes/common.hpp:34]: it could be also shared allocation. Do you handle this case?

Comment (icfaust) [reply onedal/datatypes/common.hpp:34]: Was not in the original implementation in ```onedal/datatypes/numpy/data_conversion.cpp``` but is now added to call ```wait``` on the queue to ensure data synchronization.

File: onedal/datatypes/dlpack/data_conversion.cpp (hunk 6 above)
Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp]: I realized now writing it out that since its just a pointer its not a big deal, it was previously part a function which returned a struct, so I had been worried about the copy of it, especially with respect to C++ arrays vs the C pointer used in dlpack.h. After the development, the discussion of return value optimization is a bit moot, but added it anyway. @david-cortes-intel

Comment (david-cortes-intel) [reply onedal/datatypes/dlpack/data_conversion.cpp]: I still don't get what can be assumed to be a return value optimization here, since the return value is a `py::capsule`, which behind the scenes is just a `PyObject*` (i.e. a pointer) AFAIK.

But since now the same comment appears under both if-else branches, maybe it could be moved to the beginning, or removed if it's not applicable.

Comment (icfaust) [reply onedal/datatypes/dlpack/data_conversion.cpp]: Comment has been removed.

**Review by david-cortes-intel — APPROVED** [review]
LGTM, pending CI results after merging the latest main, hopefully with this other PR included: https://github.com/uxlfoundation/scikit-learn-intelex/pull/2656

**Comment by icfaust** [comment]
Private CI failures unrelated (numpy install issue?)
